frappe/model/rename_doc.py

```diff
@@ -1,24 +1,47 @@
-# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
+# Copyright (c) 2022, Frappe Technologies Pvt. Ltd. and Contributors
 # License: MIT. See LICENSE
+from typing import TYPE_CHECKING, Dict, List, Optional
+
 import frappe
 from frappe import _, bold
 from frappe.model.dynamic_links import get_dynamic_link_map
 from frappe.model.naming import validate_name
 from frappe.model.utils.user_settings import sync_user_settings, update_user_settings_data
+from frappe.query_builder import Field
 from frappe.utils import cint
 from frappe.utils.password import rename_password
-from frappe.query_builder import Field
+
+if TYPE_CHECKING:
+	from frappe.model.meta import Meta
 
 
 @frappe.whitelist()
-def update_document_title(doctype, docname, title_field=None, old_title=None, new_title=None, new_name=None, merge=False):
+def update_document_title(
+	doctype: str,
+	docname: str,
+	title_field: Optional[str] = None,
+	old_title: Optional[str] = None,
+	new_title: Optional[str] = None,
+	new_name: Optional[str] = None,
+	merge: bool = False,
+) -> str:
 	"""
 		Update title from header in form view
 	"""
-	if docname and new_name and not docname == new_name:
+	# TODO: omit this after runtime type checking (ref: https://github.com/frappe/frappe/pull/14927)
+	for obj in [docname, new_name, new_title, old_title]:
+		if not isinstance(obj, (str, type(None))):
+			frappe.throw(f"{obj} must be of type str or None")
+
+	frappe.has_permission(doctype, ptype="write", throw=True)
```

Comment:
Should we send `docname` here as well? It is possible that only current document is permitted with `DocShare`.

Reply:
Yeah, will do. Was hesitant at first because I saw that adds an extra get_doc 😆 which might be unnecessary most of the time...but that's inconsequential compared to the seriousness of this action. Adding it 👍🏼